File: backend/src/main/java/de/interaapps/pastefy/auth/strategies/oauth2/providers/TwitchOAuth2Provider.java
package org.javawebstack.passport.strategies.oauth2.providers;

import org.javawebstack.abstractdata.AbstractObject;
import org.javawebstack.abstractdata.util.QueryString;
import org.javawebstack.http.router.util.MimeType;
import org.javawebstack.httpclient.HTTPClient;
import org.javawebstack.passport.strategies.oauth2.OAuth2Profile;
import org.javawebstack.passport.strategies.oauth2.OAuth2Provider;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

public class TwitchOAuth2Provider implements OAuth2Provider {
    private String clientId;
    private String secret;
    private String[] scopes = {"user:read:email"}; // reference: https://dev.twitch.tv/docs/authentication/#scopes

    private boolean forceVerify = false;
    private String state;

    public TwitchOAuth2Provider(String clientId, String secret){
        this.clientId = clientId;
        this.secret = secret;
    }

    public TwitchOAuth2Provider setScopes(String... scopes) {
        this.scopes = scopes;
        return this;
    }

    public OAuth2Callback callback(AbstractObject queryParameters, String callbackUrl) {
        AbstractObject abstractObject = new HTTPClient("https://id.twitch.tv").post("/oauth2/token")
                .formBodyString(new QueryString()
                        .set("client_id", clientId)
                        .set("client_secret", secret)
                        .set("code", queryParameters.string("code"))
                        .set("grant_type", "authorization_code")
                        .set("redirect_uri", callbackUrl)
                        .toString())
                .header("Accept", MimeType.JSON.getMimeTypes().get(0))
                .data().object();

        if (abstractObject.has("scope")) {
            return new OAuth2Callback(this, abstractObject.string("access_token"), abstractObject.string("refresh_token"), clientId);
        }

        return null;
    }

    public TwitchOAuth2Provider setState(String state) {
        this.state = state;
        return this;
    }

    public String getState() {
        return state;
    }

    public TwitchOAuth2Provider setForceVerify(boolean forceVerify) {
        this.forceVerify = forceVerify;
        return this;
    }

    public boolean isForceVerify() {
        return forceVerify;
    }

    public String redirect(String callbackUrl) {
        try {
            return "https://id.twitch.tv/oauth2/authorize?client_id="+clientId+"&response_type=code&scope="+ URLEncoder.encode(String.join(" ", scopes), "UTF-8")+"&redirect_uri="+URLEncoder.encode(callbackUrl, "UTF-8")+"&force_verify="+forceVerify+"&state="+state;
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static class OAuth2Callback extends org.javawebstack.passport.strategies.oauth2.OAuth2Callback {
        private String clientId;

        public OAuth2Callback(OAuth2Provider provider, String accessToken, String refreshToken, String clientId) {
            super(provider, accessToken, refreshToken);
            this.clientId = clientId;
        }

        public HTTPClient createApiClient() {
            return new HTTPClient("https://api.twitch.tv/helix").bearer(accessToken);
        }

        public OAuth2Profile getProfile() {
            OAuth2Profile profile = new OAuth2Profile();

            AbstractObject userData = createApiClient().get("/users")
                    .header("Client-Id", clientId)
                    .data().object().get("data").array().get(0).object();

            if (userData.has("id"))
                profile.id = userData.get("id").string();
            if (userData.has("login"))
                profile.name = userData.get("login").string();
            if (userData.has("profile_image_url"))
                profile.avatar = userData.get("profile_image_url").string();
            if (userData.has("email"))
                profile.mail = userData.get("email").string();
            userData.forEach(profile::set);

            return profile;
        }
    }

}

File: backend-kt/src/main/kotlin/de/interaapps/pastefy/auth/oauth/providers/TwitchOAuth2Provider.kt
package de.interaapps.pastefy.auth.oauth.providers

import de.interaapps.pastefy.auth.oauth.*

class TwitchOAuth2Provider(
    private val clientId: String = "",
    private val clientSecret: String = "",
    private val scopes: List<String> = listOf("user:read:email"),
    private val http: OAuthHttpClient? = null,
) : OAuth2Provider {
    override val name = "twitch"

    override fun authorizationUrl(callbackUrl: String, state: String): String = authorizationUrl(
        "https://id.twitch.tv/oauth2/authorize",
        mapOf("client_id" to clientId, "response_type" to "code", "scope" to scopes.joinToString(" "), "redirect_uri" to callbackUrl, "state" to state),
    )

    override fun exchangeCode(code: String, callbackUrl: String): OAuth2Tokens = requireHttp().postForm(
        "https://id.twitch.tv/oauth2/token",
        mapOf("client_id" to clientId, "client_secret" to clientSecret, "code" to code, "grant_type" to "authorization_code", "redirect_uri" to callbackUrl),
    ).tokens()

    override fun loadProfile(tokens: OAuth2Tokens): OAuth2Profile {
        val data = requireHttp().get("https://api.twitch.tv/helix/users", mapOf("Authorization" to "Bearer ${tokens.accessToken}", "Client-Id" to clientId))
            .path("data").firstOrNull() ?: throw de.interaapps.pastefy.exceptions.OAuth2Exception("Twitch returned no user")
        return OAuth2Profile(data.requiredText("id"), data.requiredText("login"), data.optionalText("email"), data.optionalText("profile_image_url"))
    }

    private fun requireHttp() = requireNotNull(http) { "OAuth HTTP client is required" }
}
